src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            var usedAssemblyLookup = usedAssemblyReferences.ToImmutableHashSet();
+
+            // Determine which toplevel references have compilation assemblies in set of used assemblies.
+            var toplevelUsedReferences = references.Where(reference =>
+                reference.CompilationAssemblies.Any(assembly => usedAssemblyLookup.Contains(assembly)));
+            var toplevelUsedAssemblyReferences = toplevelUsedReferences.SelectMany(reference => reference.GetAllCompilationAssemblies());
+
+            // Remove all assemblies that are brought into the compilation from the toplevel used references. When
+            // determining transtively used references this will reduce false positives.
+            var remainingReferences = references.Except(toplevelUsedReferences);
+            var remainingUsedAssemblyReferences = usedAssemblyReferences.Except(toplevelUsedAssemblyReferences);
+            var remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+
+            // Determine which transtive references have compilation assemblies in the set of remaining used assemblies.
+            foreach (var reference in remainingReferences)
+            {
+                var allCompilationAssemblies = reference.GetAllCompilationAssemblies();
+                if (allCompilationAssemblies.IsEmpty)
+                {
+                    // Analyzer packages won't contribute any assemblies to the compilation but will
+                    // show up as a reference. We will consider them used.
+                    continue;
+                }
+
+                if (!allCompilationAssemblies.Any(assembly => remainingUsedAssemblyLookup.Contains(assembly)))
+                {
+                    // References that do not contribute a used compilation assembly are considered unsused.
+                    unusedReferences.Add(reference);
+                    continue;
+                }
+
+                remainingUsedAssemblyReferences = remainingUsedAssemblyReferences.Except(allCompilationAssemblies);
+                remainingUsedAssemblyLookup = remainingUsedAssemblyReferences.ToImmutableHashSet();
+            }
+
+            return remainingUsedAssemblyReferences;
+        }
+
+        public async Task<Project> UpdateReferencesAsync(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates, CancellationToken cancellationToken)
+        {
+            var referenceCleanupService = project.Solution.Workspace.Services.GetRequiredService<IReferenceCleanupService>();
+
+            // Remove unused usings
+
+            foreach (var referenceUpdate in referenceUpdates)
+            {
+                if (referenceUpdate.Action == UpdateAction.None)
+                {
+                    referenceUpdate.Action = referenceUpdate.ReferenceInfo.TreatAsUsed
+                        ? UpdateAction.TreatAsUnused
```

Comment:
Should this be `TreatAsUsed`? Though I admit, I don't really understand what this is doing. If there is no reference update to apply, and the info says to treat the reference as used, then...? Perhaps a comment here would help.

Reply:
Will do